File: app/src/main/java/com/wavesplatform/wallet/v1/crypto/Hash.java
package com.wavesplatform.wallet.v1.crypto;

import fr.cryptohash.DigestEngine;
import fr.cryptohash.Keccak256;
import ove.crypto.digest.Blake2b;

public class Hash {
    public static int DigestSize = 32;

    public static byte[] hashChain(byte[] input, DigestEngine ... engines) {
        for (DigestEngine engine : engines) {
            input = engine.digest(input);
        }
        return input;
    }

    private static Blake2b.Digest blake = Blake2b.Digest.newInstance(DigestSize);
    private static Keccak256 keccak256 = new Keccak256();

    public static byte[] secureHash(byte[] input) {
        return keccak256.digest(blake.digest(input));
    }

    public static byte[] fastHash(byte[] input) {
        return blake.digest(input);
    }

}

File: wavesplatform/src/main/java/com/wavesplatform/sdk/crypto/Hash.kt
package com.wavesplatform.sdk.crypto

import com.wavesplatform.sdk.crypto.hash.Blake2b
import com.wavesplatform.sdk.crypto.hash.Keccak256
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

internal object Hash {

    private const val DIGEST_SIZE = 32

    private val blake = Blake2b.Digest.newInstance(DIGEST_SIZE)
    private val keccak256 = Keccak256()

    @JvmStatic
    fun keccak(input: ByteArray): ByteArray {
        return keccak256.digest(blake.digest(input))
    }

    @JvmStatic
    fun blake2b(input: ByteArray): ByteArray {
        return blake.digest(input)
    }

    @JvmStatic
    fun sha256(input: ByteArray): ByteArray {
        try {
            return MessageDigest.getInstance("SHA-256").digest(input)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            throw IllegalStateException("NoSuchAlgorithmException", e)
        }
    }
}
